[YashvanGH] iP

YashvanGH opened the pull request:
# Tyrese
> "Object-oriented programming is an exceptionally bad idea which could only have originated in California." 
> Edsger W. Dijkstra

### **Behold, your personal task managing assistant, Tyrese!**
With Tyrese, you can:
- ➕  Add tasks
- 🔍 Find tasks
- ❌ Delete tasks
- ✅ Mark/Unmark tasks
- ⚠️ Prioritise tasks
- 💾 Save and Load tasks

And ultimately ~~procrastinate~~ manage **ALL** your tasks efficiently!

## **📝 Task Types**
Here are some of the tasks you may want to save:
1. ✅ Todos
2. 📅 Deadlines
3. 🎟 Events

So you may ~~sometimes~~ **ALMOST ALWAYS** remember to do them on time!

## **⚙️ Features**
- [x] All the features listed above!
- [x] A basic GUI
- [ ] A nicer GUI

## **⬇️ Download**
If you're really interested, you can download tyrese [here](https://github.com/YashvanGH/ip/releases/tag/Level-10) and give it a go!

### **Note (for downloadees!)**
If you're really feeling yourself, you can download the source code, edit it and run it via the `Launcher` Class!
``` java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Tyrese.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

YashvanGH committed:
Renamed and edited Duke.java to Tyrese.java

YashvanGH committed:
Added a Greeting Class with basic functionality

YashvanGH committed:
Added add, help and list functionality

YashvanGH committed:
Added functionality to mark tasks as done and undone

YashvanGH committed:
Added ToDo, Event and Deadline Classes to differentiate different types of tasks

YashvanGH committed:
Added automated testing for all current user commands

YashvanGH committed:
Added error handling for errors such as incorrect inputs entered by the user

YashvanGH committed:
Added error handling for errors such as invalid user input

YashvanGH committed:
Added delete functionality to TaskManager Class

YashvanGH committed:
Use enums to create basic TaskPriority Class

YashvanGH committed:
Implement save and load tasks to and from hard drive

YashvanGH committed:
Merge branch-Level-7 with Level 7 features

YashvanGH committed:
Add TaskDateTimeParser Class to allow for recognition of date and time

YashvanGH committed:
Merge branch-Level-8 with Level 8 features

YashvanGH committed:
Add UserCommandParser Class to handle user input

YashvanGH committed:
Merge A-MoreOOP with OOP related features

YashvanGH committed:
Organise Classes into appropriate packages

YashvanGH committed:
Merge A-Packages with Classes in appropriate packages

YashvanGH committed:
Merge add-gradle-support into master

YashvanGH committed:
Add basic gradle automation

YashvanGH committed:
Add basic-gradle-automation branch which configured gradle

YashvanGH committed:
Add basic task tests

YashvanGH committed:
Merge basic-junit-tests branch with junit tests for tasks

YashvanGH committed:
Configure build.gradle to package app as JAR file

YashvanGH committed:
Add JavaDoc comments for all classes and methods so far

YashvanGH committed:
Merge branch-A-JavaDoc with JavaDoc comments for all classes and methods

YashvanGH committed:
Adjust code to fit coding standard

YashvanGH committed:
Merge branch-A-CodingStandard to update code to follow coding standard

YashvanGH committed:
Add find function to TaskManager Class

YashvanGH committed:
Merge branch-Level-9 with find function in TaskManager Class

YashvanGH committed:
Add GUI for Tyrese Task Manager

YashvanGH committed:
Merge branch-Level-10 with GUI features

YashvanGH committed:
Configure checkstyle functionality

YashvanGH committed:
Merge branch-A-Checkstyle with checkstyle config

wallacepck commented on src/main/java/savedata/tyrese.txt:
Perhaps the savedata folder should be added to gitignore? This would be confusing to end-users if this is built for release, unless its explicitly labelled as an example(s).

In addition, packages are not meant to be written to at runtime. I would recommend moving the save location to a folder located in the project root directory.

YashvanGH replied:
Done! And thanks for the pointer!

wallacepck commented on src/main/java/exceptions/TaskException.java:
```diff
@@ -0,0 +1,24 @@
+package exceptions;
+
+/**
+ * This class handles the exceptions thrown by operations such as mark and unmark
```
Is this summary swapped with `MarkException`?

YashvanGH replied:
Oops forgot to change the JavaDoc here 😅

wallacepck commented on src/main/java/tasks/TaskManager.java:
```diff
@@ -0,0 +1,347 @@
+        File file = new File("src/main/java/savedata/tyrese.txt");
+
+        if (!file.exists() && !file.createNewFile()) {
+            throw new IOException("You have no saved tasks! Imma initialise a new list!");
```
Is this case really an `IOException`? Perhaps a different exception is more suitable here.

YashvanGH replied:
In this sense, I feel yes because ```FileNotFound``` exception is a subclass of ```IOException```.  But I'll take note of this!

wallacepck commented on src/main/java/tasks/Deadlines.java:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author Yashvan
+ */
+public class Deadlines extends Task {
```
Perhaps `Deadlines` and `Events` should be in singular form instead? To remain consistent with `Task` and `Todo`

YashvanGH replied:
Done!

wallacepck commented on src/test/java/tasks/DeadlinesTest.java:
```diff
@@ -0,0 +1,41 @@
+
+public class DeadlinesTest {
+    @Test
+    public void testCreateValidDeadline() throws TaskException {
```
Perhaps using the given format of `featureUnderTest_testScenario_expectedBehavior()` would be more readable?

YashvanGH replied:
Will update this in a later commit!

wallacepck commented on src/main/java/tasks/TaskManager.java:
Good effort in the thorough checking of all input formatting issues, and providing the user a clear explanation of what went wrong!

Also good separation of logical units in the very long functions 👍👍👍

YashvanGH replied:
Thank you!

wallacepck commented on src/main/java/storage/SaveFileIO.java:
```diff
@@ -0,0 +1,65 @@
+ * @author Yashvan
+ */
+public class SaveFileIO {
+    private static final String DIRECTORY = "src/main/java/savedata";
```
I don't think this makes sense in release, as it would point to this directory *relative to the jar.*

wallacepck commented on src/main/java/tasks/TaskManager.java:
```diff
@@ -0,0 +1,347 @@
+     * @param input The task to find in the list.
+     */
+    public String findTasks(String input) throws TaskException {
+
```
There's an unnecessary gap here

YashvanGH replied:
Removed!

wallacepck reviewed (commented):
Overall, this follows the coding standard well; The code is descriptive, commented at important steps and formatted into a readable way. I have noted some places where refactors may be necessary.

xinnnyeee commented on src/main/java/tasks/Deadlines.java:
```diff
@@ -0,0 +1,85 @@
+        // Considered whitespace if task
+        if (parts.length < 2) {
+            throw new TaskException("PLEASE BRUH! Use: deadline <description> /by <d/M/yyyy HHmm> "
+                    + "/priority <LOW|MEDIUM|HIGH|URGENT> ._.");
```
HAHHAH love the personality

YashvanGH replied:
Haha thanks!

xinnnyeee reviewed (commented):
All coding style LGTM! Kudos to doing such a good job!

YashvanGH committed:
Refactor savedata location and rename classes

- Added \`savedata`\ folder to \`.gitignore`\ to prevent confusion.
- Moved the savedata location from package directory to project root to prevent it being written to at runtime
- Renamed \`Deadlines\` to \`Deadline\` and \`Events\` to \`Event\` for consistency.

These changes improve the maintainability of the codebase and ensure better organisation of runtime-generated files.

YashvanGH committed:
Refactor code to follow Code Quality

- Extracted out `getTimePriority` method from duplicated code segments in `Event`.
- Extracted out `extractDescription` and `extractPriority` from duplicated code segments in `TaskDateTimeParser`.
- Attempted to remove unnecessary HOW comments.
- Attempted to avoid deep nesting of code.

YashvanGH committed:
Add assertions in most files and classes

- Added assertions to check for null instances of tasklist in `Todo`, `Deadline` and `Event`.
- Added assertions to check for correct default priority instances in `Todo`, `Deadline` and `Event`.
- Added assertions in most classes to check for empty or null input and null instances of `TaskManager`.

These changes will help to ensure that the assumptions made in the program are validated during development and testing.

YashvanGH committed:
Merge pull request #1 from YashvanGH/branch-A-Assertions

Add assertions in most files and classes

YashvanGH committed:
Update .gitignore to ignore savedata directory

YashvanGH committed:
Make most classes and code follow guidelines

YashvanGH committed:
Resolve merge conflict

YashvanGH committed:
Merge pull request #2 from YashvanGH/branch-A-CodeQuality

Branch a code quality

YashvanGH committed:
Set up CI with gradle.yml file

YashvanGH committed:
Move priority task into correct package

- Contains 4 levels of priority that can be assigned to tasks (LOW, MEDIUM, HIGH, URGENT)

It is important to note that this extension was implemented as a rolling form from week 1's optional `enums` task.

YashvanGH committed:
Merge priority extension branch

YashvanGH committed:
Update workflow

YashvanGH committed:
Tweak GUI elements

YashvanGH committed:
Fix GitHub Workflow

YashvanGH committed:
Merge branch with better GUI

YashvanGH committed:
Apply some SLAP to code

YashvanGH committed:
Delete .github workflow

YashvanGH committed:
Add minor tweaks to GUI

YashvanGH committed:
Tweak README

YashvanGH committed:
Tweak README

YashvanGH committed:
Tweak README

YashvanGH committed:
Tweak README

YashvanGH committed:
Make minor changes to README

YashvanGH committed:
Attempt to apply SLAP to code

YashvanGH committed:
Add more error handling

YashvanGH committed:
Merge branch with more handling
